File: SingletonDemo/SingletonDemoTests-ObjC/GSLARCObjectTests.m
//
//  GSLARCObjectTests.m
//  SingletonDemo
//

@import Foundation;
@import XCTest;

#import "GSLARCObject.h"

@interface GSLARCObjectTests : XCTestCase
@end

@implementation GSLARCObjectTests

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testSharedARCObject {
    GSLARCObject *sharedARCObject = [GSLARCObject sharedARCObject];
    // not nil
    XCTAssertNotNil(sharedARCObject);
    // shared
    XCTAssertEqualObjects(sharedARCObject, [GSLARCObject sharedARCObject]);
    // background
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    NSInteger count = 100;
    while (count-- > 0) {
        XCTestExpectation *expectation = [self expectationWithDescription:@"background"];
        dispatch_async(queue, ^{
            XCTAssertEqualObjects(sharedARCObject, [GSLARCObject sharedARCObject]);
            [expectation fulfill];
        });
    }
    [self waitForExpectationsWithTimeout:1 handler:^(NSError *error) {
        if (error) {
            XCTFail(@"%@", error);
        }
    }];
}

- (void)testNew {
    XCTAssertNil([GSLARCObject new]);
}

- (void)testCopy {
    GSLARCObject *sharedARCObject = [GSLARCObject sharedARCObject];
    XCTAssertEqualObjects(sharedARCObject, [sharedARCObject copy]);
}

@end

File: SingletonDemo/SingletonDemoTests-Swift/GSLARCObjectTests.swift
//
//  GSLARCObjectTests.swift
//  SingletonDemo
//

import XCTest

class GSLARCObjectTests: XCTestCase {

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testShared() {
        let shared: GSLARCObject? = GSLARCObject.shared
        // not nil, just in case
        XCTAssertNotNil(shared)
        // shared
        XCTAssertEqual(GSLARCObject.shared, shared)
        // background
        for _ in 1..<100 {
            let expectation = self.expectation(description: "background")
            DispatchQueue.global(qos: .default).async {
                XCTAssertEqual(GSLARCObject.shared, shared)
                expectation.fulfill()
            }
        }
        self.waitForExpectations(timeout: 1) { errorOrNil in
            if let error = errorOrNil {
                XCTFail("\(error)")
            }
        }
    }

    func testCopy() {
        let shared = GSLARCObject.shared
        let copy = shared.copy() as? GSLARCObject
        XCTAssertEqual(copy, shared)
    }
}
